streams/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java

```diff
@@ -0,0 +1,223 @@
+import org.apache.kafka.test.TestCondition;
+import org.apache.kafka.test.TestUtils;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+@Category({IntegrationTest.class})
+public class PurgeRepartitionTopicIntegrationTest {
+
+    private static final int NUM_BROKERS = 1;
+
+    private static final String INPUT_TOPIC = "input-stream";
+    private static final String APPLICATION_ID = "restore-test";
+    private static final String REPARTITION_TOPIC = APPLICATION_ID + "-KSTREAM-AGGREGATE-STATE-STORE-0000000002-repartition";
+
+    private static AdminClient adminClient;
+    private static KafkaStreams kafkaStreams;
+    private static Integer purgeIntervalMs = 10;
+    private static Integer purgeSegmentBytes = 2000;
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(NUM_BROKERS, new Properties() {
+        {
+            put("log.retention.check.interval.ms", purgeIntervalMs);
+            put(TopicConfig.FILE_DELETE_DELAY_MS_CONFIG, 0);
+        }
+    });
+
+    private Time time = CLUSTER.time;
+
+    private class RepartitionTopicCreated implements TestCondition {
+        @Override
+        public boolean conditionMet() {
+            try {
+                Set<String> topics = adminClient.listTopics().names().get();
+
+                if (!topics.contains(REPARTITION_TOPIC)) {
+                    return false;
+                }
+            } catch (Exception e) {
+                return false;
+            }
+
+            try {
+                ConfigResource resource = new ConfigResource(ConfigResource.Type.TOPIC, REPARTITION_TOPIC);
+                Config config = adminClient.describeConfigs(Collections.singleton(resource))
+                        .values().get(resource).get();
+                return config.get(TopicConfig.CLEANUP_POLICY_CONFIG).value().equals(TopicConfig.CLEANUP_POLICY_DELETE)
+                        && config.get(TopicConfig.SEGMENT_MS_CONFIG).value().equals(purgeIntervalMs.toString())
+                        && config.get(TopicConfig.SEGMENT_BYTES_CONFIG).value().equals(purgeSegmentBytes.toString());
+            } catch (Exception e) {
+                return false;
+            }
+        }
+    }
+
+    private interface TopicSizeVerifier {
+
+        boolean verify(long currentSize);
+    }
+
+    private class RepartitionTopicPurged implements TestCondition {
+        private TopicSizeVerifier verifier;
+
+        RepartitionTopicPurged(TopicSizeVerifier verifier) {
+            this.verifier = verifier;
+        }
+
+        @Override
+        public boolean conditionMet() {
+            time.sleep(purgeIntervalMs);
```

Comment:
Why this? Couldn't this introduce a race condition? IMHO, we should advance time only after the first check that the repartitioning topic contains data?

Reply:
This is a mock time, and the Stream's committing / purging is based on that. So it would not introduce timing based race condition, but we only control it to trigger more purging.